text/0000-update_asset_packaging.md

```diff
@@ -0,0 +1,57 @@
+- Feature Name: update_asset_packaging
+- Start Date: 2016-05-10
+- RFC PR:
+- Issue:
+
+# Summary
+[summary]: #summary
+
+The current way in which we manage asset dependencies is cumbersome and leads to various issues. This RFC proposes that we replace the current method with NPM and Webpack.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, most assets (js/css) are used as gems and managed by bundler.
+A few are included directly in the vendor/ directory due to lack of gems.
+Some of the reasons that we want to change the way we currently handle this:
+
+* Some assets are not packaged as gems, so we need to either vendor or package them ourselves.
+* Some asset gems are not updated at the same rate as the upstream source.
+* Each gem dependency must be packaged separately as RPM/DEB.
+* Difficult to integrate newer JS technology such as es2015 or js modules.
+* Assets included in the vendor/ directory need manual updating when new versions are released.
+
+# Detailed design
+[design]: #detailed-design
+
+NPM will be used to manage as many of the assets that are currently being vendored or bundled as possible.
+If we run into issues with packages that are only available for Bower in the future, we will consider adding Bower as well. The main advantage of using only NPM is that we have a central location for all js dependencies and only one tool added rather then two, which reduces the complexity. NPM also allows namespacing JS modules using CommonJS to prevent possible collisions between various modules.
+The node modules will be integrated into foreman using Webpack. Webpack has a powerful system of loaders that allow handling various tasks. Webpack also comes with a small webserver that allows live-reload of all assets managed by it, hopefully leading to increased developer happiness. In addition, webpack has various other strengths that make it seem like the current leading option - ability to handle css assets and all js module flavors, smart division of code into bundles and more.
+As we will currently continue using the Rails assets pipeline, any plugin will be able to use whatever method they choose to handle any assets as long as the result is added to the pipeline (as is done currently).
+After a certain period of attempting this solution, we will review the choice and decide if we wish to continue in this direction or change to a different one.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Some reasons to stay with the current solution:
+* No extra work needed.
+* Single package management system - bundler. All dependencies are in one central location, build process is simple.
+* Rails asset pipeline doesn’t require any modifications to handle the current solution.
+
```

Comment:
Lt's examine theme plugin extension points:

1. Images: are added by modifying Rack's pipeline to serve theme's static files before serving those from the main app.
2. css: Should contain only small amounts of css rules that are necessary for adapting to the theme - they are linked at the bottom of the head element (by using deface) and are served by theme's engine.
3. Javascripts: Should contain only overrides that are relevant to the theme - they are linked at the bottom of the head element  too. The only drawback that I can think of, is introduction of modules and namespaces in the Javascript, making it more difficult to find the right method to override.

All in all, this design should not interfere with themes design, except the issue with namespaced javascript.
@tbrisker, what do you have in mind regarding this kind of issues? Do you plan adding namespaced approach to JS and if so, how would you recommend redefining methods from the main app.
Anyway I would like to see a POC that contains a theme before we can say that it's a viable option.